### PR TITLE
Implement finite volume functor argument caching

### LINKED ISSUE #20470: Implement functor caching for finite-volume-based arguments
## Reason
@joe61vette has some table-lookups that are very expensive and so we don't want to keep re-evaluating for every time we call a functor property at a given location

## Design
very similar to what we do for the finite element arguments

## Impact
speed up some finite volume computations at the expense of memory

### PR DESCRIPTION
**Pull request by lindsayad**
Got to use structured bindings and `try_emplace` for the first time.
I'm a little concerned about the cost of binary trees but I feel like
hashing of the finite volume functor arguments could be even more expensive.
I think if this shows up as an important piece of a profile, then we
can investigate an unordered_map implementation

Closes #20470

### CODE REVIEW

File: framework/include/base/MooseFunctor.h [hunk 1]
```diff
@@ -37,6 +37,12 @@ struct ElemArg
   const libMesh::Elem * elem;
   bool correct_skewness;
   bool apply_gradient_to_skewness;
+
+  friend bool operator<(const ElemArg & l, const ElemArg & r)
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/base/MooseFunctor.h:41]: It's required by the `std::less` default comparator for `std::map`. Essentially you need to define this function in order to have your map sorted

[the remark this answers is not in the document]
Comment (GiudGiud) [reply framework/include/base/MooseFunctor.h:41]: ah I expected something like this. Thanks for the info!
This PR seems ready to me

File: framework/include/base/MooseFunctor.h [hunk 2]
```diff
@@ -139,6 +153,23 @@ struct FaceArg
...
+                           l.correct_skewness,
+                           l.apply_gradient_to_skewness,
+                           l.elem_sub_id,
+                           l.neighbor_sub_id) < std::make_tuple(r.fi,
+                                                                r.limiter_type,
+                                                                r.elem_is_upwind,
+                                                                r.correct_skewness,
+                                                                r.apply_gradient_to_skewness,
+                                                                r.elem_sub_id,
+                                                                l.neighbor_sub_id);
```
Comment (GiudGiud) [inline framework/include/base/MooseFunctor.h:171]: just leaving a note that the ordering of the tuple might matter for how the map is sorted, which will matter for performance. 

Looking at this now is premature though

**Review by GiudGiud — DISMISSED** [review]

**Comment by lindsayad** [comment]
Tested the element argument caching on a single processor with a million elements (well in excess of what our users typically run on a single process) and saw that the `try_emplace` takes 2.45% of the simulation. I think this is sufficiently low to warrant not adding additional complexity at this time. I'd been considering adding `_current_elem_arg` and `_current_elem_arg_value` members (and equivalent for other argument types) and could still do it in the future.

**Review by GiudGiud — APPROVED** [review]

**Pull request merged**